AMM concept page & details on connecting to AMM-Devnet (#1616)
XRPLF/xrpl-dev-portal: content/concepts/decentralized-exchange/automated-market-makers.md

---
html: automated-market-makers.html
parent: decentralized-exchange.html
blurb: Automated Market Makers (AMMs) provide liquidity between asset pairs, complemeting the order books in the decentralized exchange while providing passive income for their liquidity providers.
status: not_enabled
labels:
  - XRP
  - Decentralized Exchange
  - AMM
---
# Automated Market Makers

> An Automated Market Maker (AMM) is a type of smart contract, built into the XRP Ledger, that provides liquidity in the decentralized exchange. An AMM holds a pool of two assets, which can be XRP or [tokens](tokens.html), and allows users to swap between them at an exchange rate set by a formula. [Offers](offers.html) and [Cross-Currency Payments](cross-currency-payments.html) automatically use a combination of Offers and AMMs to achieve cost-effective exchanges between assets.

Reviewer: ```suggestion
Automated Market Makers (AMM) are smart contracts that provide liquidity in the XRP Ledger decentralized exchange. Each AMM holds a pool of two assets and enables users to swap between them at an exchange rate set by a formula.
```
Edits for brevity.

I think talking about offers and cross-currency payments so soon muddles the understanding of AMMs. Maybe add it as a use-case so you can elaborate on how the XRP Ledger uses AMMs to facilitate?

{% include '_snippets/amm-disclaimer.md' %}

> For any given pair of assets, there can be up to one AMM in the ledger. Anyone can create an AMM for an asset pair if it doesn't exist yet, or deposit to an existing AMM's pools. Those who deposit assets into an AMM are called _liquidity providers_ (LPs) and receive "LP Tokens" from the AMM in proportion to the amounts they deposit. LP Tokens are like other fungible tokens in the XRP Ledger, except that they also give special privileges for the AMM that issues them:

Reviewer: ```suggestion
Anyone can create an AMM for a pair of assets, but it must be unique on the ledger. If an AMM already exists for a pair of assets, you can only deposit to that AMM. Those who deposit assets into an AMM are called _liquidity providers_ (LPs) and receive "LP Tokens" from the AMM. LP Tokens enable liquidity providers to:
```

> - Liquidity providers can redeem their LP Tokens to withdraw a corresponding share of the assets in the AMM's pool, including fees collected.

Reviewer: ```suggestion
- Redeem their LP Tokens for a share of the assets in the AMM pool, including fees collected.
```

> - Liquidity providers can vote to change the AMM's fee settings. The votes are weighted based on how many LP Tokens the voters hold.

Reviewer: ```suggestion
- Vote to change the AMM's fee settings. The votes are weighted based on how many LP Tokens the voters hold.
```

> - Liquidity providers can bid some of their LP Tokens to receive a temporary discount on the AMM's trading fees.

Reviewer: ```suggestion
- Bid their LP Tokens to receive a temporary discount on the AMM's trading fees.
```

When the flow of funds between the two assets in a pool is relatively active and balanced, the fees provide a source of passive income for liquidity providers. However, when the relative price between the assets shifts, the liquidity providers can take a loss on the [currency risk](https://www.investopedia.com/terms/c/currencyrisk.asp).

## How the AMM Works

> [Kris Machowski's Introduction to Automated Market Makers](https://www.machow.ski/posts/an_introduction_to_automated_market_makers/) is a good in-depth explanation on the economics and principles of AMMs in general. The XRP Ledger implements a _constant product_ AMM, which sets the exchange rate between the two assets in its pool so that multiplying the amounts of the two pools together always comes out to the same value. For example, if the AMM holds 5 ETH and 5 USD, the "constant product" value is 5×5=25. To buy 1 ETH, you'd have to spend 1.25 USD, because after removing 1 ETH from the pool and adding 1.25 USD, the constant product value is 4×6.25=25. The AMM also charges a configurable trading fee, which can be as low as 0 or as high as 1%, so you would have to pay a little more for that as well.

Reviewer: ```suggestion
The XRP Ledger implements a _constant product_ AMM, which determines the exchange rate by the ratio of assets in the pool. As the quantity of one asset goes down, the exchange rate to the other asset goes up. To see a detailed breakdown of the equation used, see: [Kris Machowski's Introduction to Automated Market Makers](https://www.machow.ski/posts/an_introduction_to_automated_market_makers/#constant-product-amms). The AMM also charges a trading fee between 0 and 1%, which is determined and redeemed by LPs.
```
Since it's a conceptual topic I'd try and keep it high-level still. The equation portion is meaty enough that they _should_ go look up how it works on the external site, but they only need to know the gist of it to grasp how the AMM works here. May want to get a second opinion on my one-liner explanation.

> **Note:** The XRP Ledger's AMM is designed to be extensible so that, with future protocol amendments, it would be possible to create an AMM where the pools are be weighted unevenly, which would make it a more general _geometric mean market maker_. However, the current implementation always uses a weight parameter of 0.5, so it functions like a constant product AMM.

Reviewer 2: Typo `are be`.

> **Note:** The XRP Ledger's AMM is designed to be extensible so that, with future protocol amendments, it would be possible to create an AMM where the pools are be weighted unevenly, which would make it a more general _geometric mean market maker_. However, the current implementation always uses a weight parameter of 0.5, so it functions like a constant product AMM.

Reviewer 2: It's not really constant, it's `sqrt(A1, A2)`.

Author: The weight parameter is always defined as 0.5.

The calculations involve exponentiation to the weight parameter—which is equivalent to square root when the weight parameter is 0.5.

So I think it's the same thing

Author: Rewrote this and other parts. Now the "How the AMM works" section ends with this:

> The XRP Ledger's implements a _geometric mean_ AMM with a weight parameter of 0.5, so it functions like a _constant product_ market maker. For a detailed explanation of the _constant product_ AMM formula and the economics of AMMs in general, see [Kris Machowski's Introduction to Automated Market Makers](https://www.machow.ski/posts/an_introduction_to_automated_market_makers/).

I think this should address the correction here without needing to discuss future possibilities (usually a bad idea to talk too much future stuff IMO) and puts the relevant details about the formulas alongside the additional reading that goes into the formula in detail.

> **Note:** The XRP Ledger's AMM is designed to be extensible so that, with future protocol amendments, it would be possible to create an AMM where the pools are be weighted unevenly, which would make it a more general _geometric mean market maker_. However, the current implementation always uses a weight parameter of 0.5, so it functions like a constant product AMM.

Reviewer: ```suggestion
**Note:** The XRP Ledger's AMM is extensible, so future amendments could introduce _geometric mean market maker_ functionality.
```

Author: Rewrote it to remove the callout and combine all the detailed discussion of formulas at the end:

> The XRP Ledger's implements a _geometric mean_ AMM with a weight parameter of 0.5, so it functions like a _constant product_ market maker. For a detailed explanation of the _constant product_ AMM formula and the economics of AMMs in general, see [Kris Machowski's Introduction to Automated Market Makers](https://www.machow.ski/posts/an_introduction_to_automated_market_makers/).

> Due to the way the constant product rule works, the bigger a proportion of the AMM's pools a trade involves, the higher the costs. This also means that, the larger the amounts in the AMM's pools, the lower the cost of making a specific trade. For example, trading 5 USD through an AMM that only holds 10 USD is more expensive than trading 5 USD through an AMM that holds 1000 USD. This is separate from the trading fee, which is always strictly a percentage of the amounts involved.

Reviewer: I think the implications of constant product rules are hard to grasp until you dig into the equations in the linked page. The topic of asset values and exchange rates gets confusing without that full context. Suggest removing this.

> An AMM can hold two assets: at most one of these can be XRP, and one or both of them can be [tokens](tokens.html). Tokens with different issuers are considered different assets for this purpose. This means that there can be an AMM for two tokens with the same currency code but different issuers ("FOO issued by WayGate" is different than "FOO issued by StableFoo"), or the same issuer but different currency codes. The order does not matter; the AMM for FOO.WayGate to XRP is the same as for XRP to FOO.WayGate.

Reviewer: I'd add this as a note above, when we talk about creating AMMs from unique asset pairs.

Author: Moved this closer to the beginning of the "How the AMM Works" section

> ## Flow of Funds

Reviewer: Maybe rename this section to LP Tokens since that seems to be the focus.

<!-- TODO: add diagrams showcasing flow of funds -->

> Whoever creates the AMM becomes the first liquidity provider, and receives LP Tokens that represent 100% ownership of assets in the AMM's pool. They can redeem some or all of those LP Tokens to withdraw assets from the AMM in proportion to the amounts currently there. (The proportions shift over time as people trade against the AMM.) The AMM does not charge a fee when withdrawing both assets.

Reviewer: The wording makes me think it might charge a fee if only withdrawing one of the assets. Is this the case?

Author: Yes, it charges a fee when withdrawing one asset. (The amount is calculated so that it's the same as if you'd done a two-asset withdrawal and then traded one of those assets for the other via the AMM.)

> For example, if you created an AMM with 5 ETH and 5 USD, then someone bought 1 ETH by spending 1.25 USD, the pool has 4 ETH and 6.25 USD in it. You can spend half your LP Tokens to withdraw 2 ETH and 3.125 USD.

Reviewer: May want to consider using the word "exchange" instead of "bought" to really drive home the idea of switching one asset for another.

Anyone can deposit assets to an existing AMM. When they do, they receive new LP Tokens based on how much they deposited. The amount that a liquidity provider can withdraw from an AMM is based on the proportion of the AMM's LP Tokens they hold compared to the total number of LP Tokens outstanding.

LP Tokens are like other tokens in the XRP Ledger, so you can use them in many [types of payments](payment-types.html), trade them in the decentralized exchange, or even deposit them as assets for new AMMs. (To receive LP Tokens as payment, you must set up a trust line with a nonzero limit with the AMM Account as the issuer.) However, you can _only_ send LP Tokens directly to the AMM (redeeming them) using the [AMMWithdraw][] transaction type, not through other types of payments. Similarly, you can only send assets to the AMM's pool through the [AMMDeposit][] transaction type.

The AMM is designed so that an AMM's asset pool is empty if and only if the AMM has no outstanding LP Tokens. This situation can only occur as the result of an [AMMWithdraw][] transaction; when it does, the AMM is automatically deleted.

> ### LP Token Currency Codes

Reviewer: Suggest rolling this into the proposed LP Tokens section.

Author: It's already a lower header level, so I'll be leaving it unchanged. It may be useful to link to this sub-section specifically.

LP Tokens use a special type of currency code that use the 160-bit hexadecimal ["non-standard" format](currency-formats.html#nonstandard-currency-codes). These codes have the first 8 bits `0x03`. The remainder of the code is a SHA-512 hash, truncated to the first 152 bits, of the two assets' currency codes and their issuers. (The assets are placed in a "canonical order" with the numerically lower currency+issuer pair first.) As a result, the LP Tokens for a given asset pair's AMM have a predictable, consistent currency code.

## Trading Fees

> The trading fee is charged on top of the implicit exchange rate between the assets in the AMM's pool. The exchange rate changes dynamically based on the balance of assets in the pool, but the trading fees only change as the result of voting. Trading fees are a source of passive income for liquidity providers, and they offset the currency risk of letting others trade against the pool's assets.

Reviewer: ```suggestion
Trading fees are charged on top of the implicit exchange rate and can only be changed by the AMM's liquidity providers. Trading fees are a source of passive income for liquidity providers, and they offset the currency risk of letting others trade against the pool's assets.
```

> Liquidity providers have an incentive to set to set trading fees at an appropriate rate: if fees are too high, trades will use order books to get a better rate instead; if fees are too low, liquidity providers don't get any benefit for contributing to the pool. <!-- STYLE_OVERRIDE: will --> Each AMM gives its liquidity providers the power to vote on its fees, in proportion to the amount of LP Tokens those liquidity providers hold.

Reviewer 2: Typo `to set to set`

Liquidity providers can vote to set the fee from 0% to 1%, in increments of 0.001%, using the [AMMVote transaction type][AMMVote]. Whenever anyone places a new vote, the AMM recalculates its fee to be an average of the latest votes weighted by how many LP Tokens those voters hold. Up to 8 liquidity providers' votes can be counted this way; if more liquidity providers try to vote then only the top 8 votes (by most LP Tokens held) are counted. Even though liquidity providers' share of LP Tokens can shift rapidly for many reasons (such as trading those tokens using [Offers](offers.html)), the trading fees are only recalculated whenever someone places a new vote (even if that vote is not one of the top 8).

### Auction Slot

> Unlike any previous Automated Market Makers, the XRP Ledger's AMM design has an _auction slot_ that a liquidity provider can bid on to get a discount on the trading fee for a 24-hour period. The bid must be paid in LP Tokens, which are returned to the AMM. No more than one account hold the auction slot at a time, but the bidder can name up to 4 additional accounts to also receive the discount. There is a minimum bid of 0.001% of the AMM's total outstanding LP Tokens, and if the slot is currently occupied then you must bid more to displace the current slot holder. If someone displaces you, you get part of your bid back depending on how much time remains. As long as you hold an active auction slot, you pay a discounted trading fee of 0% when making trades against that AMM.

Reviewer 2: `MinSlotPrice` is changed to `0`.

**Caution:** The minimum bid value is a placeholder and may change before the AMM feature becomes finalized.

> With any AMM, when the price of its assets shifts significantly in external markets shifts significantly, traders can use arbitrage to profit off the AMM, which results in a loss for liquidity providers. The auction mechanism is intended to return more of that value to liquidity providers and more quickly bring the AMM's prices back into balance with external markets.

Reviewer 2: Typo `shifts significantly` is repeated twice.
